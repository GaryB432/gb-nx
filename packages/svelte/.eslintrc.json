{
  "extends": ["../../.eslintrc.json"],
  "ignorePatterns": ["!**/*", "lib/prettier/*"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["*.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/dependency-checks": "error"
      }
    },
    {
      "files": ["./package.json", "./generators.json", "./migrations.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/nx-plugin-checks": "error"
      }
    }
  ]
}
