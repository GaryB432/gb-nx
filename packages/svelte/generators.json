{
  "$schema": "http://json-schema.org/draft-07/schema",
  "name": "svelte",
  "version": "0.0.1",
  "generators": {
    "route": {
      "factory": "./src/generators/route/generator",
      "schema": "./src/generators/route/schema.json",
      "description": "route generator"
    },
    "component": {
      "factory": "./src/generators/component/generator",
      "schema": "./src/generators/component/schema.json",
      "description": "component generator"
    },
    "dependency": {
      "factory": "./src/generators/dependency/generator",
      "schema": "./src/generators/dependency/schema.json",
      "description": "dependency generator"
    },
    "application": {
      "factory": "./src/generators/application/generator",
      "schema": "./src/generators/application/schema.json",
      "aliases": ["app"],
      "x-type": "application",
      "description": "application generator"
    },
    "init": {
      "factory": "./src/generators/init/generator",
      "schema": "./src/generators/init/schema.json",
      "description": "init generator"
    }
  }
}
