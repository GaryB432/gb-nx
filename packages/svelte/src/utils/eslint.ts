import {
  joinPathFragments,
  type ProjectConfiguration,
  type Tree,
} from '@nx/devkit';

export interface EslintConfiguration {
  env?: {
    browser: boolean;
    es2021: boolean;
    node: boolean;
  };
  extends?: string[];
  ignorePatterns?: string[];
  overrides: unknown[];
  parser?: string;
  parserOptions?: {
    ecmaVersion: string;
    extraFileExtensions: string[];
    project: string;
    sourceType: string;
  };
  plugins: string[];
  root: boolean;
  rules?: unknown;
}

function updateWorkspaceEslint(tree: Tree): void {
  const config: Partial<EslintConfiguration> = {
    root: true,
    env: {
      browser: true,
      es2021: true,
      node: true,
    },
    extends: ['plugin:gb/recommended', 'plugin:svelte/recommended'],
    overrides: [
      {
        files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
        rules: {
          '@nx/enforce-module-boundaries': [
            'error',
            {
              enforceBuildableLibDependency: true,
              allow: [],
              depConstraints: [
                {
                  sourceTag: '*',
                  onlyDependOnLibsWithTags: ['*'],
                },
              ],
            },
          ],
        },
      },
      {
        files: ['*.svelte'],
        parser: 'svelte-eslint-parser',
        parserOptions: {
          parser: '@typescript-eslint/parser',
        },
      },
    ],
    parser: '@typescript-eslint/parser',
    parserOptions: {
      ecmaVersion: 'latest',
      sourceType: 'module',
      project: 'tsconfig.base.json',
      extraFileExtensions: ['.svelte'],
    },
    plugins: ['@nx', 'gb'],
    rules: {},
  };
  tree.write('.eslintrc.json', JSON.stringify(config));
}
function updateProjectEslint(
  tree: Tree,
  project: Pick<ProjectConfiguration, 'root'>
): void {
  const config: Partial<EslintConfiguration> = {
    extends: ['../../.eslintrc.json'],
    ignorePatterns: ['.svelte-kit', 'build/*', '**/*.config.*'],
  };

  tree.write(
    joinPathFragments(project.root, '.eslintrc.json'),
    JSON.stringify(config)
  );
}

/**
 * 
 * @param tree 
 * @param project 
 * @deprecated
 */
export function updateEslint(
  tree: Tree,
  project: Pick<ProjectConfiguration, 'root'>
): void {
  updateWorkspaceEslint(tree);
  updateProjectEslint(tree, project);
}
