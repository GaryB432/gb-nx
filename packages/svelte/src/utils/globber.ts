import { parse } from 'path';

export function includes(path: string, globs: string[]): boolean {
  const parsedPath = parse(path);
  for (const c of globs) {
    if (c === path) {
      return true;
    }
    const glob = parse(c);
    if (
      glob.base === '*' &&
      glob.root === parsedPath.root &&
      glob.dir === parsedPath.dir
    ) {
      return true;
    }
  }
  return false;
}
