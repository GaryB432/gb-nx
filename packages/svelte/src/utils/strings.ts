export function stringInsert(
  source: Readonly<string>,
  pos: number,
  useComma: boolean,
  matter: string
): string {
  return [
    source.slice(0, pos),
    useComma ? ',' : '',
    matter,
    source.slice(pos),
  ].join('');
}
