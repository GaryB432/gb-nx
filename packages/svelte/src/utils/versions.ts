export const eslintPluginGbVersion = '^1.5.0';
export const eslintPluginSvelteVersion = '^2.14.1';
export const eslintVersion = '^8.28.0';
export const prettierPluginSvelteVersion = '^2.7.0';
export const typescriptEslintVersion = '^5.46.1';
