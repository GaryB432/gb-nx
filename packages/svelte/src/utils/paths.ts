import {
  joinPathFragments,
  offsetFromRoot,
  type ProjectConfiguration,
  type Tree,
} from '@nx/devkit';
import { type PackageJson } from 'nx/src/utils/package-json';

export interface NamedPath {
  name: string;
  path: string;
}

export function readRootPackageJson(tree: Tree): PackageJson | undefined {
  const pb = tree.read('package.json', 'utf-8');
  return pb ? (JSON.parse(pb) as PackageJson) : undefined;
}

export function readPackageJson(
  tree: Tree,
  config: ProjectConfiguration
): PackageJson | undefined {
  const pb = tree.read(joinPathFragments(config.root, 'package.json'));
  if (pb) {
    const pack = JSON.parse(pb.toString()) as PackageJson;
    return pack;
  }
}

export function nodeResolutionPaths(root: string): string[] {
  return root
    .split('/')
    .map((_, i, lines) => {
      return lines.slice(0, i + 1).join('/');
    })
    .reverse();
}

export function makeAliasName(name: string, scope?: string): string {
  if (!scope) {
    return name;
  }
  if (!name) {
    return '';
  }
  return ['@'.concat(scope), name].join('/');
}

export function dependencySourceRoot(
  project: ProjectConfiguration,
  dep: ProjectConfiguration
): string {
  return joinPathFragments(
    offsetFromRoot(project.root),
    dep.sourceRoot ?? joinPathFragments(dep.root, 'src')
  );
}

export function commonParentFolder(
  recordOfPaths: Record<string, unknown>
): string {
  const fdf = Object.values(recordOfPaths) as string[];

  return (
    findTopLevelCommonFolder(
      Object.values(fdf).filter((f) => typeof f === 'string')
    ) ?? 'src'
  );
}

function findTopLevelCommonFolder(folders: string[]): string | null {
  if (folders.length === 0) {
    return null;
  }

  const paths = folders.map((folder) => folder.split('/'));
  const shortestPath = paths.reduce((acc, path) =>
    path.length < acc.length ? path : acc
  );

  const commonPrefix = shortestPath.reduce((acc, segment, index) => {
    if (paths.every((path) => path[index] === segment)) {
      return acc + segment + '/';
    }
    return acc;
  }, '');

  return commonPrefix.length > 0 ? commonPrefix.slice(0, -1) : null;
}

// Tests
// const testCases: [string[], string | null][] = [
//   [["src/abc/def", "src/abc/ghi", "src/abc/fun/long", "src/abc/fun/more/names/here", "src/abc"], "src/abc"],
//   [["src/abc/def", "src/abc/ghi", "src/abc/fun/long", "src/abc/fun/more/names/here", "src/abc", "banana"], "src/abc"],
//   [["a/b/c", "a/b/d", "a/b/e"], "a/b"],
//   [["a/b", "c/d", "e/f"], null],
//   [["a"], "a"],
//   [[]], null
// ];

// testCases.forEach(([folders, expected]) => {
//   const result = findTopLevelCommonFolder(folders);
//   console.log(`Input: ${folders}, Expected: ${expected}, Result: ${result}`);
//   expect(result).toBe(expected);
// });
