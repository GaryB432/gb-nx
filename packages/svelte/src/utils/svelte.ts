import type { ProjectConfiguration, Tree } from '@nx/devkit';
import { joinPathFragments } from '@nx/devkit';
import { tsquery } from '@phenomnomnominal/tsquery';
import { type PackageJson } from 'nx/src/utils/package-json';
import { parse, valid } from 'semver';
import {
  SyntaxKind,
  type Identifier,
  type ObjectLiteralExpression,
  type PropertyAssignment,
  type SourceFile,
  type StringLiteral,
  type SyntaxList,
} from 'typescript';
import { nodeResolutionPaths, type NamedPath } from './paths';

export const SVELTE_CONFIG = 'svelte.config.js';

interface NodeApplicationSchema {
  directory: string;
  name: string;
  rootProject?: boolean;
  skipFormat?: boolean;
  skipPackageJson?: boolean;
}

interface KitFiles {
  lib: string;
  params: string;
  routes: string;
}

const defaultSvelteFiles: KitFiles = {
  lib: 'src/lib',
  params: 'src/params',
  routes: 'src/routes',
};

const SVELTE_CONFIG_KIT_OBJECT_LITERAL_SELECTOR =
  'PropertyAssignment:has(Identifier[name=kit]) ObjectLiteralExpression';

export function getKitLiteral(
  configContents: string
): ObjectLiteralExpression | undefined {
  const ast: SourceFile = tsquery.ast(configContents);
  const qresults = tsquery<ObjectLiteralExpression>(
    ast,
    SVELTE_CONFIG_KIT_OBJECT_LITERAL_SELECTOR,
    { visitAllChildren: true }
  );

  return qresults[0];
}

export function isSvelte(tree: Tree, root: string): boolean {
  return tree.exists(joinPathFragments(root, SVELTE_CONFIG));
}

function namedPathFromPropertyAssignment(node: PropertyAssignment): NamedPath {
  const identifier = node.name as Identifier;
  const name = identifier.text;

  const initializer = node.initializer as StringLiteral;
  const path = initializer.text;

  return { name, path };
}

function getSvelteFilesFromPropertyAssignment(
  filesAssignment: PropertyAssignment
): Partial<KitFiles> {
  const objectLiteral = filesAssignment.getChildren().find((node) => {
    return node.kind === SyntaxKind.ObjectLiteralExpression;
  }) as ObjectLiteralExpression | undefined;

  if (objectLiteral) {
    const syntaxList = objectLiteral.getChildren().find((node) => {
      return node.kind === SyntaxKind.SyntaxList;
    }) as SyntaxList | undefined;

    if (syntaxList) {
      const nps = syntaxList
        .getChildren()
        .filter((node) => node.kind === SyntaxKind.PropertyAssignment)
        .map<NamedPath>((n) =>
          namedPathFromPropertyAssignment(n as PropertyAssignment)
        );

      return nps.reduce(
        (a, np) => {
          a[np.name] = np.path;
          return a;
        },
        { ...defaultSvelteFiles } as Record<string, string>
      );
    }
  }

  return defaultSvelteFiles;
}

export function getSvelteFiles(configContents: string): KitFiles {
  const kitLiteral = getKitLiteral(configContents);

  if (kitLiteral) {
    const syntaxList = kitLiteral.getChildren().find((node) => {
      return node.kind === SyntaxKind.SyntaxList;
    }) as SyntaxList;

    const filesAssignment = syntaxList
      .getChildren()
      .find(
        (node) =>
          node.kind === SyntaxKind.PropertyAssignment &&
          node.getChildAt(0).getText() === 'files'
      ) as PropertyAssignment;

    if (filesAssignment) {
      return {
        ...defaultSvelteFiles,
        ...getSvelteFilesFromPropertyAssignment(filesAssignment),
      };
    }
  }

  return defaultSvelteFiles;
}

export function getSvelteConfig(
  tree: Tree,
  config: ProjectConfiguration
): string | null {
  const buf = tree.read(joinPathFragments(config.root, SVELTE_CONFIG));
  return buf ? buf.toString() : null;
}

export function getSveltePackageVersions(
  tree: Tree,
  config: ProjectConfiguration
): { name: string; version: string | undefined }[] {
  return ['@sveltejs/kit', 'svelte'].map((name) => {
    let version: string | undefined;
    for (const pjname of nodeResolutionPaths(config.root).map((p) =>
      joinPathFragments(p, 'node_modules', name, 'package.json')
    )) {
      if (tree.exists(pjname)) {
        const deppj = JSON.parse(
          tree.read(pjname, 'utf-8') ?? ''
        ) as PackageJson;
        version = deppj.version;
      }
    }
    return { name, version };
  });
}

export function satisfiesRunes(svelteDependedUpon: string): boolean {
  if (!valid(svelteDependedUpon)) {
    throw new Error('not a good semver');
  }
  const parsed = parse(svelteDependedUpon, true);
  return (parsed?.major ?? 0) > 4;
}

export function supportsRunes(
  tree: Tree,
  config: ProjectConfiguration
): { supports: boolean; svelte: string | undefined } {
  let supports = false;
  let svelte: string | undefined;
  const sps = getSveltePackageVersions(tree, config);
  const sveltePkg = sps.find((pkg) => pkg.name === 'svelte');
  // console.log(sveltePkg);
  if (sveltePkg && sveltePkg.version) {
    svelte = sveltePkg.version;
    supports = satisfiesRunes(svelte);
  }
  return { supports, svelte };
}

export function createSvelteKitApp(
  appTree: Tree,
  version: string,
  options: NodeApplicationSchema,
  devDependencies: Record<'@sveltejs/kit' | 'svelte', string> = {
    '@sveltejs/kit': '2.0.0',
    svelte: version,
  }
): void {
  const config0 = 'const config = { kit: {} };';

  const config = `/** @type {import('@sveltejs/kit').Config} */
  const config = {
    kit: {
      alias: {
        // this will match a file
        'my-file': 'path/to/my-file.js',
   
        // this will match a directory and its contents
        // (my-directory/x resolves to path/to/my-directory/x)
        'my-directory': 'path/to/my-directory',
   
        // an alias ending /* will only match
        // the contents of a directory, not the directory itself
        'my-directory/*': 'path/to/my-directory/*'
      },
      other: {
        'testing': 'yes'
      }
    }
  };`;
  const projectRoot = joinPathFragments(options.directory ?? '.', options.name);

  appTree.write(joinPathFragments(projectRoot, 'svelte.config.js'), config);

  appTree.write(
    joinPathFragments(projectRoot, 'package.json'),
    JSON.stringify({
      name: options.name,
      version: '0.0.1-tester.0',
      devDependencies,
    })
  );

  for (const [name, version] of Object.entries(devDependencies)) {
    if (!valid(version)) {
      throw new Error(`'${version}' in not a valid semver for '${name}' `);
    }

    appTree.write(
      joinPathFragments(projectRoot, 'node_modules', name, 'package.json'),
      JSON.stringify({ name, version })
    );
  }
}
