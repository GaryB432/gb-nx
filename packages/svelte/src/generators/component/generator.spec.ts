import type { Tree } from '@nx/devkit';
import { addProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { createSvelteKitApp } from '../../utils/svelte';
import componentGenerator from './generator';

import '../../utils/testing/matchers';

describe('component svelte 4', () => {
  let appTree: Tree;
  const projectName = 'my-app';

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    createSvelteKitApp(appTree, '4.0.0', {
      directory: 'apps',
      name: projectName,
      skipFormat: true,
    });
    addProjectConfiguration(appTree, projectName, { root: 'apps/my-app' });
  });

  it('should generate component in components directory', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      directory: 'components',
      skipFormat: true,
    });

    expect(
      appTree.read('apps/my-app/src/lib/components/Hello.svelte', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should generate component in lib directory by default', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      skipFormat: true,
    });

    const helloSvelte = appTree.read(
      'apps/my-app/src/lib/Hello.svelte',
      'utf-8'
    );
    expect(helloSvelte).not.toUseRunes();
    expect(helloSvelte).toContain('export let subject');
  });

  it('should generate component with scss', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      style: 'scss',
      skipFormat: true,
    });

    expect(
      appTree.read('apps/my-app/src/lib/Hello.svelte', 'utf-8')
    ).toMatchSnapshot();
  });

  it('should not generate component with runes', async () => {
    expect(async () => {
      await componentGenerator(appTree, {
        name: 'hello',
        project: projectName,
        runes: true,
        skipFormat: true,
      });
    }).rejects.toThrow(
      "runes feature requires svelte >= 5 (currently '4.0.0')"
    );
  });
});

describe('component svelte 5', () => {
  let appTree: Tree;
  const projectName = 'my-app-5';

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    createSvelteKitApp(appTree, '5.0.0', {
      directory: 'apps',
      name: projectName,
      skipFormat: true,
    });
    addProjectConfiguration(appTree, projectName, { root: 'apps/my-app-5' });
  });

  it('should generate component with runes', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      runes: true,
      skipFormat: true,
    });

    expect(
      appTree.read('apps/my-app-5/src/lib/Hello.svelte', 'utf-8')
    ).toUseRunes();
  });

  it('should default runes', async () => {
    await componentGenerator(appTree, {
      name: 'hello',
      project: projectName,
      runes: undefined,
      skipFormat: true,
    });

    expect(
      appTree.read('apps/my-app-5/src/lib/Hello.svelte', 'utf-8')
    ).toUseRunes();
  });
});
