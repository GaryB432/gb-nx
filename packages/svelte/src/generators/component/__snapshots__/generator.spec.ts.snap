// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component should generate component in components directory 1`] = `
"<script>
  
  export let subject = "Hello component";

</script>

<div class="container">
  {subject} works
</div>

<style>
  .container {
    border: thin solid silver;
  }
</style>
"
`;

exports[`component should generate component with scss 1`] = `
"<script>
  
  export let subject = "Hello component";

</script>

<div class="container">
  {subject} works
</div>

<style lang="scss">
  .container {
    border: thin solid silver;
  }
</style>
"
`;
